[clang] Constant-evaluate format strings as last resort

apple-fcloutier opened the pull request:
I [asked on the forums](https://discourse.llvm.org/t/should-attribute-format-checking-try-to-const-evaluate-strings/85854/4) and people were generally supportive of the idea, so:

Clang's -Wformat checker can see through an inconsistent set of operations. We can fall back to the recently-updated constant string evaluation infrastructure when Clang's initial evaluation fails for a second chance at figuring out what the format string is intended to be. This enables analyzing format strings that were built at compile-time with std::string and other constexpr-capable types in C++, as long as all pieces are also constexpr-visible, and a number of other patterns.

As a side effect, it also enables `tryEvaluateString` on char arrays (rather than only char pointers).

Radar-ID: rdar://99940060

apple-fcloutier committed:
[clang] Constant-evaluate format strings as last resort

Clang's -Wformat checker can see through an inconsistent set of
operations. We can fall back to the recently-updated constant string
evaluation infrastructure when Clang's initial evaluation fails for
a second chance at figuring out what the format string is intended
to be. This enables analyzing format strings that were built at
compile-time with std::string and other constexpr-capable types in
C++, as long as all pieces are also constexpr-visible, and a number
of other patterns.

Radar-ID: rdar://99940060

apple-fcloutier commented on clang/test/Sema/format-strings.c:
```diff
@@ -3,6 +3,11 @@
 // RUN: %clang_cc1 -fblocks -fsyntax-only -verify -Wformat-nonliteral -isystem %S/Inputs -triple=x86_64-linux-android %s
 
+// expected-note@-5{{format string was constant-evaluated}}
+// ^^^ there will be a <scratch space> SourceLocation caused by the
```
To clarify, in an actual diagnostic, it shows up like this:

```
<scratch space>:1:8: note: format string was constant-evaluated
    1 | "hello %s, %d"
      |        ^~
      |        %d
```

The `format string was constant-evaluated` note could say `format string was constant-evaluated to "hello %s, %d"`, or we could simply not show the constant-evaluated string. This is worse because we are unable to point at the incorrect specifier in the format string. If you had 2 or 3 of the same modifier, there would be no obvious way to tell which one we're talking about.

The patch already supports the case where constant evaluation resolves to a string literal that exists in source. When that's not the case, I feel pretty strongly that we need to bring up the format string to the user somehow to show these diagnostics. Can you think of other ways to do this?

shafik replied:
That definitely clarifies a lot, do we do this anywhere else?

apple-fcloutier replied:
Not that I'm aware of. Before this change, <scratch space> buffers are used only for macro expansion. (The other specially-named buffers are `<built-in>` and `<command line>`, but we really could call it anything.) As far as I know, diagnosing based on the string result of compile-time evaluation is unprecedented and we need to do _something_ new one way or another.

cor3ntin replied:
Like @shafik and @tbaederr - I have a slight concern with performance.
In general, we should avoid checking format strings when these diagnostics are not enabled. That would at least lead to less work in system headers. 

Benchmarking sounds like a good idea.
But I don't have a better solution than using a scratch space.

apple-fcloutier replied:
@cor3ntin I looked into this and there's about 50 distinct DiagIDs for format-related issues, which is impractical to for check ahead of time, and impractical to maintain as format diagnostics expand. Most/all of them can be controlled as an aggregate by the -Wformat and -Wformat=2 warning groups. This would be practical to check, but I think that the only facilities we have to check whether diagnostics are enabled are based on DiagIDs rather than groups.

For what it's worth, I'm less worried than you: when the format string is a function call, [we _already_ try to evaluate it](https://github.com/llvm/llvm-project/blob/16c84c4475b909d2de455a44139643c03fe3fe25/clang/lib/Sema/SemaChecking.cpp#L6225). However, the result is discarded if the lvalue base is not a string literal. This PR expands the technique as a universal fallback, but I expect the expensive case to be function calls since I think that's the only way to get control flow (aside from expression statements).

I can try to improve this to avoid having to evaluate the string twice, though.

AaronBallman replied:
I'm not too worried about performance, but I can see where the concerns come from. It may be worth it to put a branch up on https://llvm-compile-time-tracker.com/ to verify we're not slowing things down too much, but I also don't think it's strictly required. (Checking diagnostic IDs to see if the check is disabled would be really awkward and I think we should avoid it in this case.)

hubert-reinterpretcast commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -10170,6 +10170,8 @@ def warn_format_bool_as_character : Warning<
   InGroup<Format>;
 def note_format_string_defined : Note<"format string is defined here">;
+def note_format_string_evaluated_to : Note<
+  "format string was constant-evaluated">;
```
I am concerned that this wording will lead to further confusion over what constitutes "constant evaluation".

Firstly, this "constant evaluation" is not a constant evaluation required by the language (in terms of "when" constant evaluation is supposed to occur).

Secondly, as all constant evaluation required by the language occurs as-if in a manifestly constant-evaluated context, this "constant evaluation" does not match "how" constant evaluation required by the language would behave.

apple-fcloutier replied:
The point is taken. Would you like to suggest an alternate wording?

hubert-reinterpretcast replied:
> The point is taken. Would you like to suggest an alternate wording?

"format string was evaluated, for diagnostic purposes, to"?

apple-fcloutier replied:
I noodled on it and I think it's a little awkward, but I understand why we're trying to stay away from wording with a standardized definition. How do you feel about "format string resolved to a constant string"?

hubert-reinterpretcast replied:
"computed format string is"?
"format string computed to"?

apple-fcloutier replied:
I find this format awkward when it doesn't have the source immediately after. For instance, [clang-tidy displays the note without the \<scratch buffer\> contents](https://github.com/llvm/llvm-project/pull/135864#discussion_r2054861426). I like the current wording because it feels complete even when that's missing. (I know the C++ diagnostics have lots of "found this candidate", "candidate ignored because ...", "in template instantiation requested here", etc, but I think that they are a necessary evil rather than the format to strive for.)

As I understand your concerns, the main problem with my wording is that it needs an adjective to qualify "string" with that is ideally something other than "constant" because it's not used in the standard sense.

"non-literal format string evaluated/resolved/computed at compile time"? Or do you see something else in that vein that would work? There is one other diagnostic that says "compile time constant expression" (and I'm OK dropping "constant expression" for all the reasons above).

hubert-reinterpretcast replied:
> "non-literal format string evaluated/resolved/computed at compile time"

"computed format string for diagnostic purposes"?

hubert-reinterpretcast commented on clang/test/SemaCXX/format-strings.cpp:
```diff
@@ -238,3 +246,69 @@ void f(Scoped1 S1, Scoped2 S2) {
+  printf(const_string().c_str(), "hello", 123); // no-warning
+  printf(const_string().c_str(), 123, 456); // expected-warning {{format specifies type 'char *' but the argument has type 'int'}}
+}
+#endif
```
Please add a test using `if consteval` in a meaningful manner.

hubert-reinterpretcast replied:
See https://github.com/llvm/llvm-project/issues/135913 for other potential considerations.

AaronBallman replied:
FWIW, I don't see any new tests for these comments.

apple-fcloutier replied:
I added tests for a simple case, but I'm not sure what counts as "meaningful". With that said, I'm hitting the same problem that Hubert reported (which shipping Clang currently exhibits, and that my change does not address): https://godbolt.org/z/zTfGfGvKj

AaronBallman replied:
Yeah, I think the point @hubert-reinterpretcast was making is that this exacerbates an existing problem rather than introduces a new one. That's unfortunate, but perhaps we can live with it?

apple-fcloutier replied:
I played with this a bit yesterday and the last push improves the situation. When the format string checker follows a DeclRefExpr to a VarDecl that is found to have constant initialization (per `VarDecl::hasConstantInitialization`), we now honor that in the evaluation of the format string. This addresses the reduction that I made of your issue.

This does not address the issue as you reported it (where the function with `[[gnu::format]]` is constexpr/consteval itself) because at the point SemaChecking gets in play, we don't know yet that we are in a constant evaluation context. While parsing this definition:

```c++
constexpr auto myvprintf = getvprintf(const_string());
```

Format checking happens inside `Sema::ActOnCallExpr`, but we only know that `myvprintf` has constant initialization after its initializer has been parsed, and `Sema::isConstantEvaluatedContext()` is `false` at that time, so we don't have the information that we need.

I'm nervous to make changes around this because I don't understand constant evaluation well enough. I think that if we had a way to bail out of the constant evaluation machinery when running into an `if constexpr`, I could take that, but I don't think we have it either.

tbaederr commented:
I don't know how much format strings are actually exercised there, but don't forget to run this through the compile time tracker.

cor3ntin commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -5935,8 +5936,14 @@ static void CheckFormatString(
 
-static const Expr *maybeConstEvalStringLiteral(ASTContext &Context,
-                                               const Expr *E);
+enum StringLiteralConstEvalResult {
```
```suggestion
enum StringLiteralConstantEvaluationResult {
```

cor3ntin commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -5935,8 +5936,14 @@ static void CheckFormatString(
 
-static const Expr *maybeConstEvalStringLiteral(ASTContext &Context,
-                                               const Expr *E);
+enum StringLiteralConstEvalResult {
+  SLCER_NotEvaluated,
+  SLCER_NotNullTerminated,
+  SLCER_Evaluated,
+};
+
+static StringLiteralConstEvalResult
+constEvalStringAsLiteral(Sema &S, const Expr *E, const StringLiteral *&SL);
```
```suggestion
enum StringLiteralConstEvalResult {
  SLCER_NotEvaluated,
  SLCER_NotNullTerminated,
  SLCER_Evaluated,
};

static StringLiteralConstEvalResult
EvaluateStringAndCreateLiteral(Sema &S, const Expr *E, const StringLiteral *&SL);
```

cor3ntin replied:
We should really reflect we might create a scratch space here.
Also, can we comment this function?

cor3ntin commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -6331,26 +6321,79 @@ static StringLiteralCheckType checkFormatStringExpr(
-static const Expr *maybeConstEvalStringLiteral(ASTContext &Context,
-                                               const Expr *E) {
+static StringLiteralConstEvalResult
+constEvalStringAsLiteral(Sema &S, const Expr *E, const StringLiteral *&SL) {
```
```suggestion
EvaluateStringAndCreateLiteral(Sema &S, const Expr *E, const StringLiteral *&SL) {
```

cor3ntin commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -6331,26 +6321,79 @@ static StringLiteralCheckType checkFormatStringExpr(
+    {
+      llvm::raw_svector_ostream OS(EscapedString);
+      OS << '"';
+      OS.write_escaped(FormatString);
```
You should not need to do any escaping here, the diagnostics engine should take care of that for you.
You probably want tests for that.

apple-fcloutier replied:
No, I really do need the escaping, otherwise format strings containing quotes, newlines and probably other characters will print incorrectly. For instance, the format string `hello "%s"` will print as `"hello "%s""` in the scratch space when it needs to be `"hello \"%s\""`. Keep in mind that this is essentially synthesized source code, not text being piped into a diagnostic.

I don't know how to add a test for it because I don't know how to get clang -verify to surface the code in a `<scratch space>` file.

AaronBallman replied:
My big concern here is what happens to applied fix-its when the fix is in the scratch space? Do we need to suppress the fix-its in that case?

apple-fcloutier replied:
Clang uses fixits in diagnostics to show what should change, which I think is useful:

```
<scratch space>:1:8: note: format string was constant-evaluated
    1 | "hello %s, %d"
      |        ^~
      |        %d
```

`%d` here is displayed because of the fixit attached to the diagnostics.

I ran some simple tests and this is what I get:

* with `-fdiagnostics-parseable-fixits`, you do get a diagnostic entry that looks like `fix-it:"<scratch space>":{1:4-1:6}:"%d"`
* with `-Xclang -fix-what-you-can`, clang completes with no error
* with `-Xclang -fix-what-you-can -Xclang -fixit-to-temporary`, no temporary file is created in `.` or the source directory (which were different for the purposes of that test)

AaronBallman replied:
> Clang uses fixits in diagnostics to show what should change, which I think is useful:
> 
> ```
> <scratch space>:1:8: note: format string was constant-evaluated
>     1 | "hello %s, %d"
>       |        ^~
>       |        %d
> ```
> 
> `%d` here is displayed because of the fixit attached to the diagnostics.

It's useful information, but we have ways which try to automatically apply fixes and we need to make sure those behave reasonably.

> I ran some simple tests and this is what I get:
> 
>     * with `-fdiagnostics-parseable-fixits`, you do get a diagnostic entry that looks like `fix-it:"<scratch space>":{1:4-1:6}:"%d"`

That seems reasonable.

>     * with `-Xclang -fix-what-you-can`, clang completes with no error

That's good

>     * with `-Xclang -fix-what-you-can -Xclang -fixit-to-temporary`, no temporary file is created in `.` or the source directory (which were different for the purposes of that test)

I suppose that's reasonable. How about with `-fixit` which tries to apply the fix to the source file? Similar question if you run the test via `clang-tidy` and try to apply all fixes automatically.

apple-fcloutier replied:
I tried with `-fixit` and had the same result. For clang-tidy, I tried this:

```
% cat /tmp/test.c
__attribute__((format(printf, 1, 2)))
int printf(const char *, ...);

int main() {
	const char buf[] = {'"', '%', 's', '"', 0};
	printf(buf, 123);
	printf("%s", 123);
}
% bin/clang-tidy --fix /tmp/test.c
....
2 warnings generated.
/tmp/test.c:6:17: warning: format specifies type 'char *' but the argument has type 'int' [clang-diagnostic-format]
    6 |     printf(buf, 123);
      |            ~~~  ^~~
note: format string resolved to a constant string
/tmp/test.c:7:18: warning: format specifies type 'char *' but the argument has type 'int' [clang-diagnostic-format]
    7 |     printf("%s", 123);
      |             ~~   ^~~
      |             %d
/tmp/test.c:7:13: note: FIX-IT applied suggested code changes
    7 |     printf("%s", 123);
      |             ^
clang-tidy applied 1 of 1 suggested fixes.
```

In words: it fixes the `printf("%s", 123)` line to use `%d` and leaves alone the other printf alone without throwing a fuss (claiming "clang-tidy applied 1 of 1 suggested fixes"). Clang-tidy shows the "format string resolved to a constant string" note but not the scratch space contents. It's not ideal, but it's quite reasonable IMO.

AaronBallman commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -265,6 +265,9 @@ related warnings within the method body.
   ``format_matches`` accepts an example valid format string as its third
   argument. For more information, see the Clang attributes documentation.
 
+- Format string checking now supports the compile-time evaluation of format
```
It may help users to understand the improvement if there's a small code example showing what wasn't checked and is now correctly caught. WDYT?

AaronBallman commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -6331,26 +6321,79 @@ static StringLiteralCheckType checkFormatStringExpr(
+
+  // Plop that string into a scratch buffer, create a string literal and then
+  // go with that.
+  auto ScratchFile = S.getSourceManager().createFileID(std::move(MemBuf));
```
Please spell out the type.

apple-fcloutier committed:
Create StringEvalResult to hold result of compile-time string evaluation

Several users of compile-time string evaluation can meaningfully use the
special case that compile-time string evaluation resolves to a string
literal in source (for instance, to improve diagnostics). This changes
Expr::tryEvaluateString to return a StringEvalResult, which can hold
either a string literal and an offset or a std::string of evaluated
characters.

apple-fcloutier commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -17945,15 +17945,36 @@ bool Expr::tryEvaluateObjectSize(uint64_t &Result, ASTContext &Ctx,
 
 static bool EvaluateBuiltinStrLen(const Expr *E, uint64_t &Result,
                                   EvalInfo &Info, std::string *StringResult) {
-  if (!E->getType()->hasPointerRepresentation() || !E->isPRValue())
```
Mh, I don't love checking the same thing twice. I reorganized things a little differently to avoid the `else { return false }` branch.

apple-fcloutier committed:
Address naming/documentation-level review feedback

apple-fcloutier commented:
I have addressed current feedback, except that I still haven't checked for perf (will follow Aaron's directions on that soon) and I'm still working out what I need to check for `if constexpr` (confession time: I have never used it before).

apple-fcloutier commented:
(The broken tests are because I updated the note wording and then did not update the tests. We have ongoing conversation for that so I'll fix that when we have a resolution)

apple-fcloutier committed:
Reword diagnostic

AaronBallman commented on clang/test/SemaCXX/format-strings.cpp:
```diff
@@ -1,6 +1,14 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -Wformat-nonliteral -Wformat-non-iso -Wformat-pedantic -fblocks -std=c++20 %s
+
+#if __cplusplus >= 202000l
+// expected-note@-6{{format string computed from non-literal expression}}
```
Better to use `// #bookmark` and then `expected-note@#bookmark` closer to where the diagnostic comes from which generates the note. (My immediate question was "what's special about C++20?" when the answer was "the code which generates the note only runs in C++20".)

AaronBallman commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -9105,7 +9105,7 @@ def err_expected_callable_argument : Error<
   "in call to printing function with arguments '(%0)' while dumping struct">;
 def err_builtin_verbose_trap_arg : Error<
-  "argument to __builtin_verbose_trap must %select{be a pointer to a constant string|not contain $}0">;
+  "argument to __builtin_verbose_trap must %select{be a pointer to a constant NUL-terminated string|not contain $}0">;
```
```suggestion
  "argument to '__builtin_verbose_trap' must %select{be a pointer to a constant null-terminated string|not contain $}0">;
```

apple-fcloutier committed:
Aaron feedback

apple-fcloutier committed:
Getting incorrect result with consteval_str_global

ojhunt commented:
The pointer auth options PR would also benefit from this - it currently has an ad hoc and restricted version of this as it needed to support strings produced by builtins, and predated @cor3ntin's string evaluation change so that was not even remotely an option at the time.

I would much rather have us use a single string evaluation routine rather than the current behavior, but that would block the options PR on this one.

https://github.com/llvm/llvm-project/pull/136828

apple-fcloutier committed:
Fix (part of) the constexpr problem

AaronBallman commented:
I'd appreciate if you could fix the merge conflicts btw; I'd like to apply the patch locally and play around with it. The code changes look pretty reasonable to me, but I had some questions as to how stuff was handled.

apple-fcloutier commented:
Yeah, I'm fixing the merge conflict, unfortunately I got sidetracked _immediately_ after I pushed and found there was a conflict. I'll hopefully get it out by EOD.

apple-fcloutier committed:
Fix format issue

cor3ntin commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -17971,46 +18006,117 @@ static bool EvaluateBuiltinStrLen(const Expr *E, uint64_t &Result,
+std::optional<Expr::StringEvalResult>
+Expr::tryEvaluateString(ASTContext &Ctx, bool *NullTerminated,
+                        bool InConstantContext) const {
+  if (NullTerminated)
+    *NullTerminated = false;
```
Can we put that in StringEvalResult instead?

apple-fcloutier replied:
Totally, not sure what was going through my head that it wasn't there to begin with.

cor3ntin commented on clang/include/clang/AST/Expr.h:
```diff
@@ -791,10 +791,26 @@ class Expr : public ValueStmt {
-  /// constant string, return the constant string (without the terminating
-  /// null).
-  std::optional<std::string> tryEvaluateString(ASTContext &Ctx) const;
+  class StringEvalResult {
+    std::string Storage;
+    const StringLiteral *SL;
+    uint64_t Offset;
+
+  public:
+    StringEvalResult(std::string Contents);
+    StringEvalResult(const StringLiteral *SL, uint64_t Offset);
+
+    llvm::StringRef getString() const;
+    bool getStringLiteral(const StringLiteral *&SL, uint64_t &Offset) const;
+  };
```
Could that just an aggregate, that we could use with structured binding?

apple-fcloutier replied:
This is a polite `union` of `std::string` and `StringLiteral`: `Storage` is empty when `SL` is set, and vice versa. `getString()` encapsulates useful logic to decide how to get a `StringRef` back, which I think justifies having a separate type instead of an aggregate. I will add a comment on top of it.

cor3ntin commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -1920,9 +1920,17 @@ static bool EvaluateComplex(const Expr *E, ComplexValue &Res, EvalInfo &Info);
 static bool EvaluateAsRValue(EvalInfo &Info, const Expr *E, APValue &Result);
+static bool EvaluateStringAsLValue(EvalInfo &Info, const Expr *E,
+                                   QualType &CharTy, LValue &String);
+static const StringLiteral *StringLValueIsLiteral(EvalInfo &Info,
+                                                  LValue &String,
+                                                  QualType CharTy,
+                                                  uint64_t &Offset);
```
This is a pretty confusing interface.
Can we have for example?
 ````
   GetLValueAsStringLiteral(LValue) -> StringLiteral*;
   EvaluateStringLiteral(StringLiteral*, Offset) -> unsigned;
```

apple-fcloutier replied:
Happy to rename `StringLValueIsLiteral` to `GetLValueAsStringLiteral`, but I don't understand what is being asked for `EvaluateStringLiteral`. What is being evaluated and what is the `unsigned` returned value?

cor3ntin replied:
Sorry about that, I meant `EvaluateStringLiteralSize` - ie split the method in 2. Pseudo code

```cpp
if(StringLiteral* SL = GetLValueAsStringLiteral(LValue)) 
    Size =  EvaluateStringLiteralSize(SL, Offset);
```

I believe the second method never fails but I might be wrong about that, use optional if it can.
Does that make some sense?

cor3ntin commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -1920,9 +1920,17 @@ static bool EvaluateComplex(const Expr *E, ComplexValue &Res, EvalInfo &Info);
+                                                  LValue &String,
+                                                  QualType CharTy,
+                                                  uint64_t &Offset);
+template <typename CharAction>
+static bool IterateStringLValue(EvalInfo &Info, const Expr *E, QualType CharTy,
```
Can you add a comment? Maybe call it `ForEachElementInStringExpression` or something

apple-fcloutier committed:
Address cor3ntin's feedback

ojhunt commented:
@apple-fcloutier I feel like this PR can be made easier to review by separating the improvements to string constant evaluation from the improvements to the format string checking, as the former introduces a number of new interfaces and the operations required to provide those APIs, which is an independent problem from the format string checking

apple-fcloutier commented:
@AaronBallman, @cor3ntin, what do you think? Split this in two, [like Oliver suggests](https://github.com/llvm/llvm-project/pull/135864#issuecomment-2879466843)?

AaronBallman commented:
> @AaronBallman, @cor3ntin, what do you think? Split this in two, [like Oliver suggests](https://github.com/llvm/llvm-project/pull/135864#issuecomment-2879466843)?

I don't insist, but splitting may mean we can land some bits sooner while still debating other bits.
